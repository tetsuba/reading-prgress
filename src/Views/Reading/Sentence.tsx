import Button from '../../Components/Button/Button'
import Word from './Word'
import Svg from '../../Components/Svg/Svg'
import TAILWIND_CLASSES from '../../shared.tailwind'

export type WordTypes = {
    word: string
    status: string
}

type PropTypes = {
    sentence: WordTypes[]
    index: number
    count: number
    wordClickHandler?: (status: string, wordIndex: number) => void
    sentenceClickHandler?: () => void
}

export default function Sentence(props: PropTypes) {
    const completed = props.count > props.index ? 'hidden' : ''
    const active =
        props.count === props.index
            ? 'text-gray-800 delay-200'
            : 'opacity-30 text-gray-500 pointer-events-none'
    return (
        <div
            data-testid="sentence-block"
            className={`${active} ${completed} ${TAILWIND_CLASSES.sentenceBorder} relative`}
        >
            <Button
                data-testid="sentence-complete"
                template="svg"
                className="absolute right-2 bottom-1 hover:text-green-500"
                onClick={props.sentenceClickHandler}
            >
                <Svg icon="check-badge" />
            </Button>
            {props.sentence.map(({ word, status }, wordIndex) => {
                return (
                    <Word
                        key={`word-${word}-${wordIndex}`}
                        index={wordIndex}
                        status={status}
                        onClick={() => {
                            props.wordClickHandler &&
                                props.wordClickHandler(status, wordIndex)
                        }}
                    >
                        {word}
                    </Word>
                )
            })}
        </div>
    )
}
