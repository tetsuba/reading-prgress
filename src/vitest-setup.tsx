import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from './store/store'
import { updateUser } from './store/user/userSlice'
import { addBooks } from './store/books/booksSlice'
import { addStudents } from './store/students/studentsSlice'
import { studentsMockData } from '../tests/mockData/students'
import booksMockData from '../tests/mockData/books'
import { userMockData } from '../tests/mockData/user'

const portal = document.createElement('div')
portal.setAttribute('id', 'modal')
document.body.appendChild(portal)

const mockResults = {
    results: [[{ transcript: 'This is a story' }]]
}

class MockSpeech {
    start() {
        // @ts-ignore
        this.onresult(mockResults)
    }
    stop = vi.fn()
    onresult = vi.fn()
    onend = vi.fn()
}

Object.defineProperty(window, 'webkitSpeechRecognition', {
    writable: true,
    value: vi.fn().mockImplementation(() => new MockSpeech())
})

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            staleTime: 0,
            cacheTime: 0,
            retryDelay: 1,
            retry: 0
        }
    }
})

type PropTypes = {
    pathname: string
    children: JSX.Element
    notAuthenticated?: boolean
}

export function updateStoreWithMockData() {
    store.dispatch(updateUser(userMockData))
    store.dispatch(addBooks(booksMockData))
    store.dispatch(addStudents(studentsMockData))
}

export function WrapperWith_Store_Query_Router(props: PropTypes) {
    if (!props.notAuthenticated) updateStoreWithMockData()
    return (
        <Provider store={store}>
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[props.pathname]}>
                    {props.children}
                </MemoryRouter>
            </QueryClientProvider>
        </Provider>
    )
}

export function WrapperWith_Store_Router(props: PropTypes) {
    return (
        <Provider store={store}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </Provider>
    )
}

export function WrapperWithQueryAndRouter(props: PropTypes) {
    return (
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </QueryClientProvider>
    )
}

export function WrapperWithQuery(props: { children: JSX.Element }) {
    return (
        <QueryClientProvider client={queryClient}>
            {props.children}
        </QueryClientProvider>
    )
}

export function WrapperWithRouter(props: PropTypes) {
    return (
        <MemoryRouter initialEntries={[props.pathname]}>
            {props.children}
        </MemoryRouter>
    )
}
