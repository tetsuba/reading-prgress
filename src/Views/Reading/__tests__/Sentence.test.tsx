import {render} from "@testing-library/react";
import {WrapperWithRouter} from "../../../vitest-setup";
import Sentence from "../Sentence";

describe('Sentence', () => {
    test('if props data is undefined', () => {
        const { asFragment } = render(
            <WrapperWithRouter pathname="/">
                <Sentence data={undefined} index={0} count={0} sentenceClickHandler={() => {}} wordClickHandler={() => {}}/>
            </WrapperWithRouter>
        )
        expect(asFragment()).toMatchSnapshot()
    })
})