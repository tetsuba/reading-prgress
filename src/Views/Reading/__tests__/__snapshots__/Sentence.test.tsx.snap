// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sentence > if props data is undefined 1`] = `
<DocumentFragment>
  <div
    class="text-gray-800 delay-200 min-h-96 border-dashed md:border-x-2 border-gray-200 border-y-2 md:rounded-lg p-4 text-2xl md:p-6 mb-8 relative"
    data-testid="sentence-block"
  >
    <button
      class="bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400 hover:text-green-500 absolute bottom-1 right-2"
      data-testid="sentence-complete"
    >
      <svg
        class="h-6 w-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 12.75L11.25 15 15 9.75M21 12c0 1.268-.63 2.39-1.593 3.068a3.745 3.745 0 01-1.043 3.296 3.745 3.745 0 01-3.296 1.043A3.745 3.745 0 0112 21c-1.268 0-2.39-.63-3.068-1.593a3.746 3.746 0 01-3.296-1.043 3.745 3.745 0 01-1.043-3.296A3.745 3.745 0 013 12c0-1.268.63-2.39 1.593-3.068a3.745 3.745 0 011.043-3.296 3.746 3.746 0 013.296-1.043A3.746 3.746 0 0112 3c1.268 0 2.39.63 3.068 1.593a3.746 3.746 0 013.296 1.043 3.746 3.746 0 011.043 3.296A3.745 3.745 0 0121 12z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
